include/vikunja/mem/iterator/ConstantIterator.hpp

```diff
@@ -0,0 +1,267 @@
+                {
+                    assert(v == other.v && "Can't subtract constant iterators of different values!");
+                    return ConstantIterator(v, index - other.index);
+                }
+
+                /**
+                 * @brief Add an index to this iterator
+                 */
+                NODISCARD ALPAKA_FN_INLINE ConstantIterator& operator+=(const IdxType idx)
+                {
+                    index += idx;
+                    return *this;
+                }
+
+                /**
+                 * @brief Subtract an index from this iterator
+                 */
+                NODISCARD ALPAKA_FN_INLINE ConstantIterator& operator-=(const IdxType idx)
+                {
+                    index -= idx;
+                    return *this;
+                }
+
+#pragma endregion arithmeticoperators
+
+#pragma region comparisonoperators
+
+// if spaceship operator is available is being used we can use spaceship operator magic
+#ifdef USESPACESHIP
+
+                /**
+                 * @brief Spaceship operator for comparisons
+                 */
+                NODISCARD ALPAKA_FN_INLINE auto operator<=>(const ConstantIterator& other) const noexcept = default;
+
+// if cpp20 *isn't* defined we get to write 70 lines of boilerplate
+#else
+
+                /**
+                 * @brief Equality comparison, returns true if the iterators are the same
+                 */
+                NODISCARD ALPAKA_FN_INLINE bool operator==(const ConstantIterator& other) const noexcept
+                {
+                    return v == other.v && index == other.index;
+                }
+
+                /**
+                 * @brief Inequality comparison, negated equality operator
+                 */
+                NODISCARD ALPAKA_FN_INLINE bool operator!=(const ConstantIterator& other) const noexcept
+                {
+                    return !operator==(other);
+                }
+
+                /**
+                 * @brief Less than comparison, value is checked first, then index
+                 */
+                NODISCARD ALPAKA_FN_INLINE bool operator<(const ConstantIterator& other) const noexcept
+                {
+                    if(v < other.v)
+                        return true;
+                    if(v > other.v)
+                        return false;
+                    return index < other.index;
+                }
+
+                /**
+                 * @brief Greater than comparison, value is checked first, then index
+                 */
+                NODISCARD ALPAKA_FN_INLINE bool operator>(const ConstantIterator& other) const noexcept
+                {
+                    if(v > other.v)
+                        return true;
+                    if(v < other.v)
+                        return false;
+                    return index > other.index;
+                }
```

Comment:
When implementing comparison operators, you usually only implement `operator<` and `operator==`. The rest should be based on those:

```suggestion
                NODISCARD ALPAKA_FN_INLINE bool operator>(const ConstantIterator& other) const noexcept
                {
                    return other < *this;
                }
```

This way you can avoid accidentially only defining a partial order for your type. Or inconsistent comparions (e.g.  `a < b` and `a > b` at the same time).

Applies to other operators below.

Reply:
I know this can be done and reduces the probability to make mistakes, but I was trying to be explicit so I was sure the compiler wouldn't add unnecessary function calls. But I can change it anyways, if you think that this is unnecessary

Reply:
Compiler inlining heuristics are quite good! I am certain, the compiler would inline the delegation to `operator<` inside `operator>`.